go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,193 @@
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			return nil, fmt.Errorf("error adding EgressFirewall DNS rule for host %s, in namespace %s: addressSetFactory is nil", dnsName, namespace)
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		e.signalAdded(dnsName)
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			e.dns.Delete(dnsName)
+			// the dnsEntry is no longer needed because nothing references it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) updateEntryForName(dnsName string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	ips := e.dns.GetIPs(dnsName)
+	e.dnsEntries[dnsName].dnsResolves = ips
+	if err := e.dnsEntries[dnsName].dnsAddressSet.SetIPs(ips); err != nil {
+		return fmt.Errorf("cannot add IPs from EgressFirewall AddressSet %s: %v", dnsName, err)
+	}
+	return nil
+}
+
+// Run spawns a goroutine that handles updates to the dns entries for dnsNames used in
+// EgressFirewalls. The loop runs after receiving one of two signals
+// 1. a new dnsName has been added and a signal is sent to add the new DNS name, if an
+//    EgressFirewall uses a DNS name already added by another egressFirewall the previous
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsName string
+	var ttl time.Time
+	var timeSet bool
+	// initially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsName := <-e.added:
+				if err := e.dns.Add(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
```

Comment:
I think this is not doing what you expect :) https://stackoverflow.com/a/39212626

Reply:
@aojea just as the person in the article you linked to I defined the timer outside the loop so the right behaviour is seen. I have a test case to exercise this specific scenario 

https://github.com/ovn-org/ovn-kubernetes/pull/1597/files#diff-f485ca3ff4cc809c2b4d05ccd8172e3d893410abbeee7c8871ac8bb3660b4e9eR169

Reply:
yeah, but 
> Each time you execute time.After(4 * time.Second) you create a new timer channel.

each time that you executes `time.After(durationTillNextQuery)` you have a new channel, so it works because you eventually runs time.After(0).

Reply:
@aojea I think he wants the duration to refresh if he got a dnsName added right? So the code seems OK. If we wanted to use a hardcoded timer that didnt change we could use ticker.

Reply:
yes the timer would need to be re-evaluated when a dnsName gets added

Reply:
ok now i see what you mean @aojea its a little confusing but it works I guess. I think it would be better to split updating ttl expire dns and adding new dns into separate goroutines. Can do that in a follow up PR if desired.

Reply:
actually I I'm wrong and this works as expected, my apologies.
This is blocked in the `select .. case` so the time.After is only executed once, my mind was assuming that it was going out of the select case statement and renewing the timer, this works https://play.golang.org/p/duR0bZOVWl2
Please discard my comment @JacobTanenbaum